core/src/main/java/org/apache/calcite/rel/rules/ProjectJoinJoinRemoveRule.java

```diff
@@ -0,0 +1,144 @@
+ *
+ * <p>becomes
+ *
+ * <blockquote>
+ * <pre>select s.product_id, pc.product_id from
+ * sales as s
+ * left join product_class pc
+ * on s.product_id = pc.product_id</pre></blockquote>
+ *
+ */
+public class ProjectJoinJoinRemoveRule extends RelOptRule {
+  public static final ProjectJoinJoinRemoveRule INSTANCE
+      = new ProjectJoinJoinRemoveRule(LogicalProject.class,
+      LogicalJoin.class, RelFactories.LOGICAL_BUILDER);
+
+  /** Creates a ProjectJoinJoinRemoveRule. */
+  public ProjectJoinJoinRemoveRule(
+      Class<? extends Project> projectClass,
+      Class<? extends Join> joinClass, RelBuilderFactory relBuilderFactory) {
+    super(
+        operand(projectClass,
+            operandJ(joinClass, null,
+                join -> join.getJoinType() == JoinRelType.LEFT,
+                operandJ(joinClass, null,
+                    join -> join.getJoinType() == JoinRelType.LEFT, any()))),
+        relBuilderFactory, null);
+  }
+
+  @Override public void onMatch(RelOptRuleCall call) {
+    final Project project = call.rel(0);
+    final Join topJoin = call.rel(1);
+    final Join bottomJoin = call.rel(2);
+    int leftBottomChildSize = bottomJoin.getLeft().getRowType().getFieldCount();
+
+    // Check whether the project uses unexpected columns.
+    if (!RelOptUtil.removeJoin(project, leftBottomChildSize,
+        bottomJoin.getRowType().getFieldCount())) {
+      return;
+    }
+
+    // Check whether the top join uses unexpected columns.
+    final List<Integer> leftKeys = new ArrayList<>();
+    RelOptUtil.splitJoinCondition(topJoin.getLeft(), topJoin.getRight(),
+        topJoin.getCondition(), leftKeys, new ArrayList<>(),
+        new ArrayList<>());
+    if (leftKeys.stream().filter(s -> s >= leftBottomChildSize).count() > 0) {
+      return;
+    }
+
+    // Check whether left join keys in top join and bottom join are equal.
+    final List<Integer> leftChildKeys = new ArrayList<>();
+    final List<Integer> rightChildKeys = new ArrayList<>();
+    RelOptUtil.splitJoinCondition(bottomJoin.getLeft(), bottomJoin.getRight(),
+        bottomJoin.getCondition(), leftChildKeys, rightChildKeys,
+        new ArrayList<>());
+    if (!leftKeys.equals(leftChildKeys)) {
+      return;
+    }
+
+    // Make sure that right keys of bottom join are unique.
+    final ImmutableBitSet.Builder columns = ImmutableBitSet.builder();
+    rightChildKeys.forEach(key -> columns.set(key));
+    final RelMetadataQuery mq = call.getMetadataQuery();
+    if (!mq.areColumnsUnique(bottomJoin.getRight(), columns.build())) {
+      return;
+    }
+
+    int offset = bottomJoin.getRight().getRowType().getFieldCount();
+    final RexBuilder rexBuilder = topJoin.getCluster().getRexBuilder();
+
+    RexShuttle shuttle = new RexShuttle() {
+      @Override public RexNode visitInputRef(RexInputRef inputRef) {
+        return rexBuilder.makeInputRef(inputRef.getType(),
+            inputRef.getIndex() - offset);
+      }
+    };
```

Comment:
I think we have some utility functions that can do the shuttle work. 
RexUtil.shift?

Reply:
Good point. Will check it.